### PR TITLE
moving genericapiserver command line flags to genericapiserver

### PR DESCRIPTION
**Pull request by nikhiljindal**
cc @kubernetes/sig-api-machinery

### CODE REVIEW

File: pkg/util/flag/flags.go [hunk 1]
```diff
@@ -47,5 +49,31 @@ func WarnWordSepNormalizeFunc(f *pflag.FlagSet, name string) pflag.NormalizedNam
 func InitFlags() {
 	pflag.CommandLine.SetNormalizeFunc(WordSepNormalizeFunc)
 	pflag.CommandLine.AddGoFlagSet(goflag.CommandLine)
+	ParseFlags()
+}
+
+func ParseFlags() {
```
Comment (lavalamp) [inline pkg/util/flag/flags.go]: Why is this needed now but not previously?

Comment (nikhiljindal) [inline pkg/util/flag/flags.go]: We did not need this previously since ParseFlags() was not being called from a test before.
I added a call to ParseFlags() in Run() in examples/apiserver/apiserver.go which is also being called from examples/apiserver/apiserver_test.go 

Another way to "fix" this is to not call ParseFlags from Run() and instead call it in main in which case, ServerOptions related logic will need to be moved to main. And we will not be able to test flags parsing code.

**Comment by nikhiljindal** [comment]
@k8s-bot e2e test this: https://github.com/kubernetes/kubernetes/issues/24148

**Comment by nikhiljindal** [comment]
@k8s-bot deleted my previous comment!

@k8s-bot e2e test this issue: https://github.com/kubernetes/kubernetes/issues/24148

File: examples/apiserver/apiserver.go [hunk 2]
```diff
@@ -93,6 +96,9 @@ func Run() error {
 		return fmt.Errorf("Error in installing API: %v", err)
 	}
 	serverOptions := genericapiserver.NewServerRunOptions()
+	serverOptions.AddFlags(pflag.CommandLine)
+	flag.InitFlags()
```
Comment (lavalamp) [inline examples/apiserver/apiserver.go]: IMO, constructing options should happen in a step before Run()

File: pkg/util/flag/flags.go (hunk 1 above)
Comment (lavalamp) [inline pkg/util/flag/flags.go]: IMO we should not be testing flag parsing in a way that uses the test's invocation's own arguments, that seems quite broken.

Comment (nikhiljindal) [inline pkg/util/flag/flags.go]: hmm. Updated the code to run InitFlags only in main.go.
Not calling it in test

File: examples/apiserver/apiserver.go (hunk 2 above)
Comment (nikhiljindal) [inline examples/apiserver/apiserver.go]: Updated this function to take ServerRunOptions as input argument.

**Comment by nikhiljindal** [comment]
Updated the code as per comments. PTAL

File: examples/apiserver/server/main.go [hunk 3]
```diff
@@ -18,12 +18,21 @@ package main
...
 	"k8s.io/kubernetes/examples/apiserver"
+	"k8s.io/kubernetes/pkg/util/flag"
 
 	"github.com/golang/glog"
+	"github.com/spf13/pflag"
 )
 
 func main() {
-	if err := apiserver.Run(); err != nil {
+
```
Comment (lavalamp) [inline examples/apiserver/server/main.go]: nit: extra blank line

**Comment by lavalamp** [comment]
This looks tons better, thanks. LGTM (add label after you fix the nit)

File: examples/apiserver/server/main.go (hunk 3 above)
Comment (nikhiljindal) [inline examples/apiserver/server/main.go]: removed

**Comment by nikhiljindal** [comment]
Updated the code as per comment. Adding the LGTM tag.
